core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantPartitioningScheme.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import io.trino.metadata.InternalNode;
+import io.trino.metadata.Split;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.ToIntFunction;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class FaultTolerantPartitioningScheme
+{
+    private final int partitionCount;
+    private final Optional<int[]> bucketToPartitionMap;
+    private final Optional<ToIntFunction<Split>> splitToBucketFunction;
+    private final Optional<List<InternalNode>> partitionToNodeMap;
+
+    @VisibleForTesting
+    FaultTolerantPartitioningScheme(
+            int partitionCount,
+            Optional<int[]> bucketToPartitionMap,
+            Optional<ToIntFunction<Split>> splitToBucketFunction,
+            Optional<List<InternalNode>> partitionToNodeMap)
+    {
+        checkArgument(partitionCount > 0, "partitionCount must be greater than zero");
+        this.partitionCount = partitionCount;
+        this.bucketToPartitionMap = requireNonNull(bucketToPartitionMap, "bucketToPartitionMap is null");
+        this.splitToBucketFunction = requireNonNull(splitToBucketFunction, "splitToBucketFunction is null");
+        this.partitionToNodeMap = requireNonNull(partitionToNodeMap, "splitToBucketFunction is null").map(ImmutableList::copyOf);
```

Comment:
should you check that there is an entry for each partition in `partitionToNodeMap`?